Fix crashes from the `git-history` and `git-commit-detail` widgets

vince-fugnitto opened the pull request:
Fixes #1770
Fixes #1754

- Fixes error thrown when attempting to perform `Date.toLocaleDateString` in the `git-commit-detail-widget`.
- Fixes error thrown when attempting to iterate over `fileChangeNodes` and it is `undefined`.

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

vince-fugnitto commented:
@kittaakos sure I'll make the changes. Just to clarify, when searching for `moment` I only find it for the `blame-decorator` under the `@theia/git` extension. Is this what you're referring to?

kittaakos commented:
> Is this what you're referring to?

Yes.

CC: @AlexTugarev

vince-fugnitto commented:
@kittaakos from what I gather, `moment` is used to display a clear representation of the commit's date relative to the user's current date, (ex: **5 months ago**). If we were to switch to plain ISO strings, then we'd have to re-implement a lot of the methods and logic already present in `moment`. Is this still something worthwhile?

I noticed `moment` is also being used in `@theia/workspace`, so perhaps if we really want to implement our own functions we should do it in a separate PR that addresses both packages together.

![selection_011](https://user-images.githubusercontent.com/40359487/48068901-08990e00-e1a2-11e8-95cf-27244ba47502.png)

AlexTugarev commented:
> I think we have to get rid of moment

@kittaakos, is it because of the size of the lib? in that case it's be worth a try to use a version without locale support. maybe in [webpack](https://www.npmjs.com/package/moment-locales-webpack-plugin)? 

OTOH in case of the blame view, I think it's very convenient to see the age of a change. I guess it's easier for users to understand the heat map, right?

<img width="178" alt="screen shot 2018-11-06 at 15 52 52" src="https://user-images.githubusercontent.com/914497/48072126-1d929300-e1dc-11e8-80c9-92d25162d6c3.png">
vs.

<img width="229" alt="screen shot 2018-11-06 at 15 52 33" src="https://user-images.githubusercontent.com/914497/48072129-1e2b2980-e1dc-11e8-8caa-20dbae9b43e4.png">

👍 to use ISO everywhere! it works fine with momentjs.

vince-fugnitto commented:
@kittaakos just for my own knowledge, how would I go about updating our `Date` to use `ISO` instead of `epoch` time and would would the benefits be?

kittaakos commented:
I am not sure if I got this correctly, but if you are asking what to change, then you have to replace [`%ad`](https://github.com/theia-ide/theia/blob/1efad7cb520eb760d75e40c24e97e6b6adce4897/packages/git/src/node/dugite-git.ts#L114) with the strict [ISO 8601 format](http://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.15); [`%aI`](https://git-scm.com/docs/git-log). Probably change the [parser](https://github.com/theia-ide/theia/blob/1efad7cb520eb760d75e40c24e97e6b6adce4897/packages/git/src/node/dugite-git.ts#L145) and the [model](https://github.com/theia-ide/theia/blob/1efad7cb520eb760d75e40c24e97e6b6adce4897/packages/git/src/common/git-model.ts#L332) too. The frontend code has to be adjusted too.

Why? More details [here](https://github.com/theia-ide/theia/issues/1754#issuecomment-385390341).

I hope this helps. Does it make sense? If no, let me know.

vince-fugnitto commented:
> I hope this helps. Does it make sense? If no, let me know.

Thanks @kittaakos, I'll look into the comments you mentioned, at first glance it wasn't obvious to me on what had to be done to refactor to using ISO.

kittaakos commented:
Please note, 

https://github.com/theia-ide/theia/blob/f37496b9956dc4039156a8262a278e0f67738a42/packages/git/src/common/git-model.ts#L337

might become obsolete as the ISO string contains the timezone:
```
Z	is the time zone offset specified as “Z” (for UTC) or either “+” or “-” followed by a time expression HH:mm
```

vince-fugnitto commented:
> Please note,
> 
> [theia/packages/git/src/common/git-model.ts](https://github.com/theia-ide/theia/blob/f37496b9956dc4039156a8262a278e0f67738a42/packages/git/src/common/git-model.ts#L337)
> 
> Line 337 in [f37496b](/theia-ide/theia/commit/f37496b9956dc4039156a8262a278e0f67738a42)
> 
>  readonly tzOffset?: number; 
> might become obsolete as the ISO string contains the timezone:
> 
> ```
> Z	is the time zone offset specified as “Z” (for UTC) or either “+” or “-” followed by a time expression HH:mm
> ```

thanks!

vince-fugnitto commented:
@kittaakos I've updated the PR to use ISO strings as you requested :) let me know if there are any issues, I wasn't 100% sure on what was required from the `pumpEntry` parser so I followed other examples.

kittaakos commented:
Why do we need to change the [`git-history-widget.tsx`](https://github.com/theia-ide/theia/pull/3410/files#diff-f151e9af88b4bffb05b2872de3e3e8ef)?

vince-fugnitto commented:
> Why do we need to change the [`git-history-widget.tsx`](https://github.com/theia-ide/theia/pull/3410/files#diff-f151e9af88b4bffb05b2872de3e3e8ef)?

I was fixing an error from thrown by `bad revision` and found the code for `addCommits` was quite confusing, so I simplified it and removed the unnecessary timeout call. If you'd prefer I could rollback those changes. I do however find its a lot easier now to understand and simplifies the code.

kittaakos commented on packages/git/src/browser/history/git-history-widget.tsx:
```diff
@@ -258,7 +254,7 @@ export class GitHistoryWidget extends GitNavigableListWidget<GitHistoryListNode>
             content = <div className='message-container'>
                 <div className='no-history-message'>
-                    <div>There is no Git history available{path}.</div>
+                    <div>There is no git history available{path}.</div>
```
`git` -> `Git`

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -141,14 +141,12 @@ export class CommitDetailsParser extends OutputParser<CommitWithChanges> {
-            const [sha, email, name, timeAsString, authorDateRelative, summary, body, rawChanges] = chunk.trim().split(CommitDetailsParser.ENTRY_DELIMITER);
-            const timestamp = parseInt(timeAsString, 10);
+            const [sha, email, name, timestamp, authorDateRelative, summary, body, rawChanges] = chunk.trim().split(CommitDetailsParser.ENTRY_DELIMITER);
+            // const timestamp = parseInt(timeAsString, 10);
```
Remove.

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -141,14 +141,12 @@ export class CommitDetailsParser extends OutputParser<CommitWithChanges> {
-                author: {
-                    timestamp, email, name
-                },
+                author: { timestamp, email, name },
```
It's not required.

kittaakos replied:
Never mind. Leave as is but please next time, do not change if not necessary. It does not add much value to the code and it is easier to review.

kittaakos commented:
> I do however find its a lot easier now to understand and simplifies the code.

@jbicker, please take a look into this PR.

kittaakos commented on packages/git/src/browser/history/git-history-widget.tsx:
```diff
@@ -124,18 +124,15 @@ export class GitHistoryWidget extends GitNavigableListWidget<GitHistoryListNode>
-            });
-            log.then(async changes => {
+            try {
+                let changes = await this.git.log(repository, options);
```
The previous state did not have the `await`. Are you OK with the change, @jbicker?

jbicker replied:
I'll have a look now.

jbicker replied:
Ok, works as before. I am ok with the change as it actually simplifies the code. 
Thanks you, @vince-fugnitto!

kittaakos replied:
👍

vince-fugnitto replied:
thanks @jbicker !

kittaakos commented:
The [commit date](https://github.com/theia-ide/theia/issues/1754) is correct:

![screen shot 2018-11-08 at 11 41 41](https://user-images.githubusercontent.com/1405703/48193676-4dac7400-e34b-11e8-861e-ea4c9a3c78d7.jpg)
![screen shot 2018-11-08 at 11 40 05](https://user-images.githubusercontent.com/1405703/48193677-4dac7400-e34b-11e8-895d-e08e1a4b0c4d.jpg)

And I can refresh the browser, [no more errors due to the commit details](https://github.com/theia-ide/theia/issues/1770).

Thanks for the fix @vince-fugnitto!

vince-fugnitto commented:
thanks @kittaakos, would you also prefer if I now remove the `authorTz` from the code ? It looks like I forgot to do so.

kittaakos commented:
> if I now remove the `authorTz` from `dugite-git` ?

From where? I thought you switched from the epoch to the ISO String, but we still need it for the blame as a property.

vince-fugnitto commented:
> > if I now remove the `authorTz` from `dugite-git` ?
> 
> From where? I thought you switched from the epoch to the ISO String, but we still need it for the blame as a property.

I was just wondering if it had to be removed as well but if not that's fine.

https://github.com/theia-ide/theia/blob/1ecc750e7096798199cb539ab6f1257fddcffa8d/packages/git/src/node/dugite-git.ts#L241

https://github.com/theia-ide/theia/blob/1ecc750e7096798199cb539ab6f1257fddcffa8d/packages/git/src/node/dugite-git.ts#L268

vince-fugnitto commented:
if its fine @kittaakos do you mind merging? I am not yet a committer :(

kittaakos commented:
The failing Windows test is unrelated: https://github.com/theia-ide/theia/issues/2586.

I am merging this.

kittaakos commented:
> you mind merging?

Done. I did not delete the branch; please do so if you do not need it.

vince-fugnitto commented:
thanks @kittaakos for your help and reviews!
